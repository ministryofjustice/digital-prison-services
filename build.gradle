task wrapper(type: Wrapper) {
    gradleVersion = '4.6'
}





