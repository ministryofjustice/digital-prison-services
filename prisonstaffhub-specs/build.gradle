plugins {
    id "groovy"
}

ext {

    ext {
        drivers = [
//                'chrome'
//                ,
                'chromeHeadless'
        ]

        groovyVersion = '2.4.14'
        gebVersion = '2.1'
        seleniumVersion = '3.11.0'
        chromeDriverVersion = '2.37'
    }
}

repositories {
    mavenCentral()
}

dependencies {
    testCompile('org.spockframework:spock-core:1.1-groovy-2.4') {
        exclude group: "org.codehaus.groovy"
    }

    testCompile "org.codehaus.groovy:groovy-all:$groovyVersion"
    testCompile "org.gebish:geb-core:$gebVersion"
    testCompile "org.gebish:geb-spock:$gebVersion"
    testCompile "org.seleniumhq.selenium:selenium-support:$seleniumVersion"
    testCompile "org.seleniumhq.selenium:selenium-chrome-driver:$seleniumVersion"
    testCompile 'io.github.http-builder-ng:http-builder-ng-apache:1.0.3'

    testCompile 'com.github.tomakehurst:wiremock-standalone:2.15.0'
    testCompile 'com.github.tomjankes:wiremock-groovy:0.2.0'
    testCompile 'org.slf4j:slf4j-api:1.7.25'
    testCompile 'ch.qos.logback:logback-classic:1.2.3'
}

drivers.each { driver ->
    task "${driver}Test"(type: Test) {
        group JavaBasePlugin.VERIFICATION_GROUP

        outputs.upToDateWhen { false } // Always run tests

        systemProperty 'geb.build.reportsDir', reporting.file("geb/${name}")
        systemProperty 'geb.env', driver
    }
}

test {
    dependsOn drivers.collect { tasks["${it}Test"] }
    enabled = false
}

