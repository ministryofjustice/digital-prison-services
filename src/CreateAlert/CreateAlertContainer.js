import React, { Component } from 'react'
import PropTypes from 'prop-types'
import axios from 'axios'
import { connect } from 'react-redux'
import { notify } from 'react-notify-toast'
import GridRow from '@govuk-react/grid-row'
import GridCol from '@govuk-react/grid-col'

import OffenderPage from '../OffenderPage/OffenderPage'
import CreateAlertForm from './CreateAlertForm'

export class CreateAlertContainer extends Component {
  constructor(props) {
    super(props)
    this.state = {
      alertTypes: [],
      alertSubTypes: [],
    }

    this.createAlertHandler = this.createAlertHandler.bind(this)
    this.cancel = this.cancel.bind(this)
  }

  async componentDidMount() {
    const response = await axios.get('/api/get-alert-types')
    this.setState(state => ({
      ...state,
      ...response.data,
    }))
  }

  async createAlertHandler(values) {
    const { offenderDetails, handleError, history, raiseAnalyticsEvent, resetErrorDispatch } = this.props
    const { bookingId } = offenderDetails
    const { alertType, alertSubType, comment, effectiveDate } = values

    resetErrorDispatch()
    try {
      await axios.post(`/api/create-alert/${bookingId}`, {
        alertType,
        alertSubType,
        comment,
        effectiveDate,
      })
      raiseAnalyticsEvent({
        category: 'alert created',
        label: 'Alerts',
        value: { alertType, alertSubType, comment, effectiveDate },
      })
      notify.show('Alert has been created', 'success')
      history.goBack()
    } catch (error) {
      handleError(error)
    }
  }

  cancel() {
    const { history } = this.props

    history.goBack()
  }

  render() {
    const { offenderNo, handleError, setLoadedDispatch, offenderDetails } = this.props
    const { alertTypes, alertSubTypes } = this.state
    return (
      <>
        <GridRow>
          <GridCol columnOneHalf>
            <OffenderPage
              alwaysRender
              title={() => `Create alert`}
              handleError={handleError}
              offenderNumber={offenderNo}
              setLoaded={setLoadedDispatch}
            />
          </GridCol>
        </GridRow>
        <GridRow>
          <GridCol>
            <p className="label margin-bottom-small">Name</p>
            <p>
              <strong>
                {offenderDetails.firstName} {offenderDetails.lastName} ({offenderNo})
              </strong>
            </p>
          </GridCol>
        </GridRow>
        <GridRow>
          <GridCol columnOneHalf>
            <CreateAlertForm
              cancelHandler={this.cancel}
              createAlertHandler={this.createAlertHandler}
              alertTypes={alertTypes}
              alertSubTypes={alertSubTypes}
            />
          </GridCol>
        </GridRow>
      </>
    )
  }
}

CreateAlertContainer.propTypes = {
  offenderNo: PropTypes.string.isRequired,
  offenderDetails: PropTypes.shape({
    firstName: PropTypes.string,
    lastName: PropTypes.string,
  }).isRequired,
  handleError: PropTypes.func.isRequired,
  raiseAnalyticsEvent: PropTypes.func.isRequired,
  history: PropTypes.shape({
    replace: PropTypes.func.isRequired,
  }).isRequired,
  resetErrorDispatch: PropTypes.func.isRequired,
  setLoadedDispatch: PropTypes.func.isRequired,
}

const mapStateToProps = state => ({
  error: state.app.error,
  offenderDetails: state.offenderDetails,
  activeCaseLoadId: state.app.user.activeCaseLoadId,
})

const mapDispatchToProps = {}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(CreateAlertContainer)
