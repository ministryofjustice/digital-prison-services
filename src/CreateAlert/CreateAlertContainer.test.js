import React from 'react'
import { shallow } from 'enzyme/build'
import axios from 'axios'
import MockAdapter from 'axios-mock-adapter'
import { notify } from 'react-notify-toast'
import { CreateAlertContainer } from './CreateAlertContainer'

const axiosMock = new MockAdapter(axios)

const apiCalls = {
  getAlertTypes: '/api/get-alert-types',
  createAlert: '/api/create-alert/1',
}

const defaultProps = {
  offenderNo: 'A12345',
  offenderDetails: {
    bookingId: 1,
    firstName: 'John',
    lastName: 'Doe',
  },
  alertTypes: [],
  alertSubTypes: [],
  resetErrorDispatch: () => {},
  setLoadedDispatch: () => {},
  handleError: () => {},
  createAlertHandler: () => {},
  raiseAnalyticsEvent: () => {},
  history: {
    replace: () => {},
    goBack: () => {},
  },
}

const alerts = {
  alertTypes: [{ title: 'alert 1', value: 'alert-type-1' }],
  alertSubTypes: [{ title: 'alert sub 1', value: 'alert-sub-1', parentValue: 'alert-type-1' }],
}

const findGetRequests = url => axiosMock.history.get.find(request => request.url === url)

describe('Create Alert container', () => {
  beforeEach(() => {
    axiosMock.reset()
    axiosMock.onGet(apiCalls.getAlertTypes).reply(200, alerts)
    notify.show = jest.fn()
  })

  describe('Initial load', () => {
    it('should request alert types and sub types', async () => {
      const wrapper = shallow(<CreateAlertContainer {...defaultProps} />)
      const instance = wrapper.instance()

      await instance.componentDidMount()

      const matching = findGetRequests(apiCalls.getAlertTypes)
      expect(matching).toBeDefined()
    })

    it('should update the state with new alert types and sub types', async () => {
      const wrapper = shallow(<CreateAlertContainer {...defaultProps} />)
      const instance = wrapper.instance()

      await instance.componentDidMount()

      expect(wrapper.state()).toEqual(alerts)
    })

    it('should pass the correct props to the CreateForm component', async () => {
      const tree = shallow(<CreateAlertContainer {...defaultProps} {...alerts} />)

      expect(tree).toMatchSnapshot()
    })
  })

  describe('Creating an alert', () => {
    it('should create new alert', async () => {
      axiosMock.onPost(apiCalls.createAlert).reply(201)

      const wrapper = shallow(<CreateAlertContainer {...defaultProps} {...alerts} />)
      const instance = wrapper.instance()
      await instance.createAlertHandler({
        alertType: 'P',
        comment: 'sdf',
        alertSubType: 'PL1',
        effectiveDate: '2019-10-10',
      })

      const postRequest = axiosMock.history.post[0]

      expect(postRequest.url).toBe(apiCalls.createAlert)
      expect(JSON.parse(postRequest.data)).toEqual({
        alertType: 'P',
        comment: 'sdf',
        alertSubType: 'PL1',
        effectiveDate: '2019-10-10',
      })
    })

    it('should handle error from failed create alert attempt', async () => {
      axiosMock.onPost(apiCalls.createAlert).reply(500)

      const handleError = jest.fn()

      const wrapper = shallow(<CreateAlertContainer {...defaultProps} handleError={handleError} />)
      const instance = wrapper.instance()
      await instance.createAlertHandler({ alertType: 'P', comment: 'sdf', alertSubType: 'PL1' })

      expect(notify.show.mock.calls.length).toBe(0)
      expect(handleError).toHaveBeenCalledWith(new Error('Request failed with status code 500'))
    })

    it('should show success notification and redirect back to previous page on create alert', async () => {
      axiosMock.onPost(apiCalls.createAlert).reply(201)
      const history = {
        goBack: jest.fn(),
        replace: () => {},
      }

      const wrapper = shallow(<CreateAlertContainer {...defaultProps} {...alerts} history={history} />)
      const instance = wrapper.instance()
      await instance.createAlertHandler({ alertType: 'P', comment: 'sdf', alertSubType: 'PL1' })

      expect(notify.show).toHaveBeenCalledWith('Alert has been created', 'success')
      expect(history.goBack).toHaveBeenCalled()
    })

    it('should raise analytics event', async () => {
      axiosMock.onPost(apiCalls.createAlert).reply(201)
      const raiseAnalyticsEvent = jest.fn()

      const wrapper = shallow(
        <CreateAlertContainer {...defaultProps} {...alerts} raiseAnalyticsEvent={raiseAnalyticsEvent} />
      )
      const instance = wrapper.instance()
      await instance.createAlertHandler({
        alertType: 'P',
        comment: 'sdf',
        alertSubType: 'PL1',
        effectiveDate: '2019-10-10',
      })

      expect(raiseAnalyticsEvent).toHaveBeenCalledWith({
        category: 'alert created',
        label: 'Alerts',
        value: { alertType: 'P', comment: 'sdf', alertSubType: 'PL1', effectiveDate: '2019-10-10' },
      })
    })

    it('should reset errors before creating a new alert', async () => {
      axiosMock.onPost(apiCalls.createAlert).reply(201)
      const resetErrorDispatch = jest.fn()

      const wrapper = shallow(
        <CreateAlertContainer {...defaultProps} {...alerts} resetErrorDispatch={resetErrorDispatch} />
      )
      const instance = wrapper.instance()
      await instance.createAlertHandler({ alertType: 'P', comment: 'sdf', alertSubType: 'PL1' })

      expect(resetErrorDispatch).toHaveBeenCalled()
    })

    it('should navigate back to the previous location on cancel', async () => {
      const history = {
        goBack: jest.fn(),
        replace: () => {},
      }
      const wrapper = shallow(<CreateAlertContainer {...defaultProps} {...alerts} history={history} />)
      const instance = wrapper.instance()

      instance.cancel()

      expect(history.goBack).toHaveBeenCalled()
    })
  })
})
