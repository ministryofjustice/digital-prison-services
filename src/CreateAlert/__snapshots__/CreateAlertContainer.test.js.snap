// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create Alert container Initial load should pass the correct props to the CreateForm component 1`] = `
<Fragment>
  <GridRow>
    <GridCol
      columnFull={false}
      columnOneHalf={false}
      columnOneQuarter={false}
      columnOneThird={false}
      columnThreeQuarters={false}
      columnTwoThirds={false}
      setWidth="one-half"
    >
      <Connect(OffenderPage)
        alwaysRender={true}
        handleError={[Function]}
        offenderNumber="A12345"
        setLoaded={[Function]}
        title={[Function]}
      />
    </GridCol>
  </GridRow>
  <GridRow>
    <GridCol
      columnFull={false}
      columnOneHalf={false}
      columnOneQuarter={false}
      columnOneThird={false}
      columnThreeQuarters={false}
      columnTwoThirds={false}
    >
      <p
        className="label margin-bottom-small"
      >
        Name
      </p>
      <p>
        <strong>
          John
           
          Doe
           (
          A12345
          )
        </strong>
      </p>
    </GridCol>
  </GridRow>
  <GridRow>
    <GridCol
      columnFull={false}
      columnOneHalf={false}
      columnOneQuarter={false}
      columnOneThird={false}
      columnThreeQuarters={false}
      columnTwoThirds={false}
      setWidth="one-half"
    >
      <Connect(CreateAlertFinalForm)
        alertSubTypes={Array []}
        alertTypes={Array []}
        cancelHandler={[Function]}
        createAlertHandler={[Function]}
        disableSubmit={false}
      />
    </GridCol>
  </GridRow>
</Fragment>
`;
